### ISSUE TITLE
--loop/--loop-file stutters between loops on --gpu-api=vulkan

### ISSUE THREAD

**Issue by Rabcor** [mpv-player/mpv#13924]
### Important Information

Provide following Information:
- mpv version
```
mpv v0.38.0 Copyright © 2000-2024 mpv/MPlayer/mplayer2 projects
 built on Apr 18 2024 06:31:17
libplacebo version: v6.338.2
FFmpeg version: n6.1.1
FFmpeg library versions:
   libavutil       58.29.100
   libavcodec      60.31.102
   libavformat     60.16.100
   libswscale      7.5.100
   libavfilter     9.12.100
   libswresample   4.12.100
```
- Linux Distribution and Version: EndeavourOS, kernel 6.6.28-1-lts
- Source of the mpv binary: Official Repo
- Window Manager and version: Hyprland (I believe I had this issue on X too though though I can't confirm it atm)
- Tested on both AMD and Nvidia GPU
-

### Reproduction steps

mpv --no-config --gpu-api=vulkan --loop perfect-loop.gif

It is easiest to see the stutter on perfectly looping low-framerate short (~1-2s) animations. It does not have to be any particular format (e.g. it's the same if it's mp4 or gif)

### Expected behavior

Seamless looping

### Actual behavior

Stutters at the start/end of each loop. I believe it might be skipping the first or last frame.

Possible relation to this issue: #2745

### Log file

[output.txt](https://github.com/mpv-player/mpv/files/15025920/output.txt)

### Sample files
The issue is fairly easy to see with this GIF.
![5xtDary2ULj8lEba812](https://github.com/mpv-player/mpv/assets/5684325/f6a96aee-8715-4705-a1b3-7d9b48eee1e5)

**Comment by guidocella** [comment]
Probably fixed by https://github.com/mpv-player/mpv/pull/10748

**Comment by Dudemanguy** [comment]
I don't see any stutter on my end.

**Comment by Rabcor** [comment]
> I don't see any stutter on my end.

:thinking: interesting, are you on X11 or Wayland?

**Comment by Dudemanguy** [comment]
I tried both and I don't see anything. But maybe I'm just blind.

**Comment by Rabcor** [comment]
> I tried both and I don't see anything. But maybe I'm just blind.

I just tried on X11 too, it stuttered there but I feel like it was less obvious than on Wayland.

it is just 1 frame tho I think, but to me it's blatantly obvious, did you try playing it side-by-side with the opengl renderer on one instance (or maybe even just your browser)? it might be easier to spot the difference if you have a perfectly smooth sample beside it.

**Comment by kasper93** [comment]
Are you able to test with #10748 changes?

**Comment by Rabcor** [comment]
> Are you able to test with #10748 changes?

I wouldn't be averse to trying, but I really don't know how, I can barely manage to wrangle pkgbuilds into building specific versions from git for me if I have like 20 minutes to do it, never mind applying unapplied pull requests.

Also, this file can serve as an alternate sample, it's bigger and higher framerate so it's a little easier to notice with this file.

Might be worth a try for you to use this file instead @Dudemanguy 

https://giphy.com/gifs/daily-after-effects-hypnotic-oO95KMJ7LqZxdh66pY

With this file I noticed there actually is a stutter in the opengl renderer between loops as well, but if the vulkan renderer stutter is a normal stutter, the one in the opengl renderer is a microstutter (had to really be looking for it to see it)

**Comment by guidocella** [comment]
I also can't reproduce the stutter on either video on either Vulkan or a 2011 laptop.

**Comment by kasper93** [comment]
> if I had a prepared patch that's compatible with the current git master I could manage it... But that PR is from 2022, so I doubt it'll be very easy to get that.

According to github there are no conflicts, you can use this patch directly https://github.com/mpv-player/mpv/pull/10748.patch

**Comment by Dudemanguy** [comment]
>Might be worth a try for you to use this file instead @Dudemanguy
>https://giphy.com/gifs/daily-after-effects-hypnotic-oO95KMJ7LqZxdh66pY

I can see it here I think. opengl looks fine to me.

#10748 seems to completely break the duration of gifs at least.

**Comment by Rabcor** [comment]
> > if I had a prepared patch that's compatible with the current git master I could manage it... But that PR is from 2022, so I doubt it'll be very easy to get that.
> 
> According to github there are no conflicts, you can use this patch directly https://github.com/mpv-player/mpv/pull/10748.patch

Yep, getting same result as Dudemanguy. It fixes the stutter, but it breaks the duration of gifs, although i think that is pretty much how it work maybe (instead of starting the file over, it keeps playing the file in a continuous stream).

I also took a look to see if there was a performance difference, performance didn't change at all from using the patch.

Honestly if the duration thingy is fixed, that code might be good to use as default for --loop/--loop-file, using a --seamless-looping setting seems like nonsense.

**Comment by Rabcor** [comment]
I added a recording to the OP.

**Comment by Rabcor** [comment]
I added a recording to the op of the effect

**Comment by Dudemanguy** [comment]
That has nothing to do with this.

**Comment by Rabcor** [comment]
@Dudemanguy I have disovered something interesting, it might be a clue to how it might be possible to get this running.

If I use this autoloop script https://github.com/zc62/mpv-scripts/blob/master/autoloop.lua which automatically enables looping for small files, and I run them with `mpv --vf=loop=-1 file.mp4`

The first loop will be seamless, but the subsequent loops will hitch like in my examples.

However, if I do not use the script and just do  `mpv --vf=loop=-1 --loop-file=inf file.mp4` it hitches even on the first run.

I think that `vf=loop-1` must be set after loop-file is enabled, not before. And I think maybe if the filter gets reapplied on every loop it won't ever hitch, though I have not tested that theory, i will in a minute though.

**Comment by Rabcor** [comment]
I have created a seamless looping script based on this:

> Setting --vf=loop=-1:size=289 Where 289 is the exact number of frames guarantees seamless video looping

You can get it here

https://github.com/Rabcor/mpv-scriptdump/blob/main/seamlessloop.lua

It's an ugly workaround but it works about as well as that PR https://github.com/mpv-player/mpv/pull/10748 used to. Easier to use even because it uses mp.observe to watch the loop setting so whenever loop is enabled the seamless looping logic kicks in automatically.

**Comment by Rabcor** [comment]
Not sure what PR it was, but this is now solved in the main branch.
